File: coding/Honeypot/htb/solver.py
#!/usr/bin/env python3
from collections import defaultdict, deque

def solve():
  n = int(input())
  # adjencency list instead of Node class - faster
  tree = defaultdict(list)
  for _ in range(n-1):
    p, c = map(int, input().split(" - "))
    tree[p].append(c)
    tree[c].append(p)
  honeypot = int(input().strip())

  parent = [0] * (n + 1)
  contains_honeypot = [False] * (n + 1)
  subtree_size = [0] * (n + 1)

  post_order = []
  stack = deque([1])
  while stack:
    node = stack.pop()
    post_order.append(node)
    for neighbor in tree[node]:
      if neighbor == parent[node]: continue
      parent[neighbor] = node
      stack.append(neighbor)

  for node in reversed(post_order):
    subtree_size[node] = 1
    contains_honeypot[node] = node == honeypot
    for neighbor in tree[node]:
      if neighbor == parent[node]: continue
      subtree_size[node] += subtree_size[neighbor]
      contains_honeypot[node] |= contains_honeypot[neighbor]
  
  biggest_subtree, firewall_node = 0, 0
  for node in range(1, n + 1):
    if subtree_size[node] > biggest_subtree and not contains_honeypot[node]:
      biggest_subtree = subtree_size[node]
      firewall_node = node
  
  print(n - biggest_subtree)

solve()

File: coding/Honeypot/htb/solver.cc
#include <iostream>
#include <vector>
#include <unordered_map>
#include <stack>
#include <sstream>
#include <cstdlib>
#include <ctime>
#include <thread>
#include <chrono>
#include <string>

int main() {
    int n;
    std::cin >> n;
    std::cin.ignore(); // consume newline

    std::unordered_map<int, std::vector<int>> tree;

    // Read "a - b" formatted edges
    for (int i = 0; i < n - 1; ++i) {
        std::string line;
        std::getline(std::cin, line);
        std::stringstream ss(line);
        int a, b;
        char dash;
        ss >> a >> dash >> b;
        tree[a].push_back(b);
        tree[b].push_back(a);
    }

    int honeypot;
    std::cin >> honeypot;

    std::vector<int> parent(n + 1, 0);
    std::vector<bool> contains_honeypot(n + 1, false);
    std::vector<int> subtree_size(n + 1, 0);
    std::vector<int> post_order;

    std::stack<int> s;
    s.push(1);
    while (!s.empty()) {
        int node = s.top();
        s.pop();
        post_order.push_back(node);
        for (int neighbor : tree[node]) {
            if (neighbor == parent[node]) continue;
            parent[neighbor] = node;
            s.push(neighbor);
        }
    }

    for (int i = static_cast<int>(post_order.size()) - 1; i >= 0; --i) {
        int node = post_order[i];
        subtree_size[node] = 1;
        contains_honeypot[node] = (node == honeypot);
        for (int neighbor : tree[node]) {
            if (neighbor == parent[node]) continue;
            subtree_size[node] += subtree_size[neighbor];
            contains_honeypot[node] = contains_honeypot[node] || contains_honeypot[neighbor];
        }
    }

    int biggest_subtree = 0;
    for (int node = 1; node <= n; ++node) {
        if (!contains_honeypot[node] && subtree_size[node] > biggest_subtree) {
            biggest_subtree = subtree_size[node];
        }
    }

    std::cout << (n - biggest_subtree) << std::endl;

    return 0;
}
